File: ros2_self_msg_srv/src/my_interfaces_cpp/src/subscriber_person.cpp
/**
 * 订阅自定义消息示例 - C++版本
 */
#include "rclcpp/rclcpp.hpp"
#include "my_interfaces/msg/person.hpp"

class SubscriberPerson : public rclcpp::Node
{
public:
    SubscriberPerson() : Node("subscriber_person")
    {
        // 创建订阅者，订阅自定义消息Person
        subscription_ = this->create_subscription<my_interfaces::msg::Person>(
            "person_info", 10,
            std::bind(&SubscriberPerson::topic_callback, this, std::placeholders::_1));
        
        RCLCPP_INFO(this->get_logger(), "订阅者已启动，等待消息...");
    }

private:
    void topic_callback(const my_interfaces::msg::Person::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), 
            "收到: 姓名=%s, 年龄=%d, 身高=%.2fm",
            msg->name.c_str(), msg->age, msg->height);
    }

    rclcpp::Subscription<my_interfaces::msg::Person>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SubscriberPerson>());
    rclcpp::shutdown();
    return 0;
}

File: ros2_self_msg_srv/src/my_interfaces_py/my_interfaces_py/subscriber_person.py
#!/usr/bin/env python3
"""订阅自定义消息示例 - Python版本"""
import rclpy
from rclpy.node import Node
from my_interfaces.msg import Person


class SubscriberPerson(Node):
    """订阅人员信息节点"""

    def __init__(self):
        super().__init__('subscriber_person')
        # 创建订阅者，订阅自定义消息Person
        self.subscription = self.create_subscription(
            Person,
            'person_info',
            self.listener_callback,
            10
        )
        self.subscription  # 防止未使用警告
        
        self.get_logger().info('订阅者已启动，等待消息...')

    def listener_callback(self, msg):
        """消息回调：处理接收到的消息"""
        self.get_logger().info(
            f'收到: 姓名={msg.name}, 年龄={msg.age}, 身高={msg.height:.2f}m'
        )


def main(args=None):
    rclpy.init(args=args)
    subscriber = SubscriberPerson()
    try:
        rclpy.spin(subscriber)
    except KeyboardInterrupt:
        pass
    finally:
        subscriber.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
